README.md
# cmvision

## cmvision port for ROS Melodic. Tested on Ubuntu 18.04

Steps to get it working:

0.) Clone this repository into your workspace.

1.) This package depends on wxWidgets. Download the source file for wxWidgets from this [link](https://www.wxwidgets.org/downloads/).

2.) Follow these [instructions](https://wiki.wxwidgets.org/Compiling_and_getting_started) to compile and install the package correctly. You can use this [link](https://www.binarytides.com/install-wxwidgets-ubuntu/) as an additional guide but I wouldn't entirely depend on it.

3.) I ran into an issue with the FindwxWidgets.cmake file while compiling the cmvision package with catkin. This file can be located at: /usr/share/cmake-3.10/Modules
I had to use the modified file provided [here](https://github.com/microsoft/vcpkg/issues/4756#issuecomment-620122321). Replace the contents of FindwxWidgets.cmake file with these modifications.

4.) After doing a sudo apt-get update && sudo apt-get upgrade you should be good to go. Now compile cmvision using catkin_make.

5.) In the catkin_ws/src/cmvision/cmvision.launch file, set the camera topic to your camera topic and in the catkin_ws/src/cmvision/colors.txt specify the blob colors you 
need to detect.

6.) After running roslaunch cmvision cmvision.launch, you should be able to see the results visually and by using rostopic echo /blobs


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package cmvision
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Forthcoming
-----------
* Merge pull request `#2 <https://github.com/utexas-bwi/cmvision/issues/2>`_ from OSUrobotics/hydro-devel
  Include color name in the blob message.
* Contributors: Alex Hubers

0.4.1 (2013-09-03)
------------------
* fixed missing fltk dependency. closes `#1 <https://github.com/utexas-bwi/cmvision/issues/1>`_

0.4.0 (2013-08-30)
------------------
* Ported cmvision to github from  wg-ros-pkg
* Catkinized for release into ROS Hydro


src/cmvision.h
/*=========================================================================
    CMVision.h
  -------------------------------------------------------------------------
    API definition for the CMVision real time Color Machine Vision library
  -------------------------------------------------------------------------
    Copyright 1999, 2000         #### ### ### ## ## ## #### ##  ###  ##  ##
    James R. Bruce              ##    ####### ## ## ## ##   ## ## ## ######
    School of Computer Science  ##    ## # ## ## ## ##  ### ## ## ## ## ###
    Carnegie Mellon University   #### ##   ##  ###  ## #### ##  ###  ##  ##
  -------------------------------------------------------------------------
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
  =========================================================================*/

#ifndef __CMVISION_H__
#define __CMVISION_H__

// uncomment if your compiler supports the "restrict" keyword
// #define restrict __restrict__
#define restrict

#include <stdlib.h>
#include <string.h>
#include <stdio.h>

/*
Ultra-fast intro to processing steps:
 - Color segmentation
   - load / save
   - set new values
   - process frame
 - Connected Regions
   - RLE
   - merge
   - extract blobs
   - separate blobs
   - sort blobs
   - merge blobs
 - Blob merging (not currently in release)
   - merge by area occupied

Options File Format: (RGB merge name)
[Colors]
(00,00,00) 0.95 'Orange'
(00,00,00) 0.00 'Pink'
(00,00,00) 0.00 'Red'
(00,00,00) 0.00 'DarkBlue'
(00,00,00) 0.00 'Blue'

[Thresholds]
(<lo>:<hi>,<lo>:<hi>,<lo>:<hi>)
(<lo>:<hi>,<lo>:<hi>,<lo>:<hi>)
(<lo>:<hi>,<lo>:<hi>,<lo>:<hi>)
*/

#define CMV_COLOR_LEVELS  256
#define CMV_MAX_COLORS     32

// sets tweaked optimal values for image size
#define CMV_DEFAULT_WIDTH  320
#define CMV_DEFAULT_HEIGHT 240

// values may need tweaked, although these seem to work usually
#define CMV_MAX_RUNS     (CMV_DEFAULT_WIDTH * CMV_DEFAULT_HEIGHT) / 4
#define CMV_MAX_REGIONS  CMV_MAX_RUNS / 4
#define CMV_MIN_AREA     20

#define CMV_NONE ((unsigned)(-1))

#ifndef NULL
#define NULL (0)
#endif

struct lab{
  unsigned char l,a,b;
};

typedef struct lab image_pixel;

#ifndef RGB_STRUCT
#define RGB_STRUCT
struct rgb{
  unsigned char red,green,blue;
};
#endif

// Options for level of processing
//   use enable()/disable() to change 
#define CMV_THRESHOLD      0x01
#define CMV_COLOR_AVERAGES 0x02
#define CMV_DUAL_THRESHOLD 0x04
#define CMV_DENSITY_MERGE  0x08

#define CMV_VALID_OPTIONS  0x0F


class CMVision{
public:
  struct region{
    int color;          // id of the color
    int area;           // occupied area in pixels
    int x1,y1,x2,y2;    // bounding box (x1,y1) - (x2,y2)
    float cen_x,cen_y;  // centroid
    lab average;        // average color (if CMV_COLOR_AVERAGES enabled)

    int sum_x,sum_y,sum_z; // temporaries for centroid and avg color
    // int area_check; // DEBUG ONLY

    region *next;       // next region in list

    // int number; // DEBUG ONLY
  };

  struct rle{
    unsigned color;     // which color(s) this run represents
    int length;         // the length of the run (in pixels)
    int parent;         // run's parent in the connected components tree
  };

  struct color_info{
    rgb color;          // example color (such as used in test output)
    char *name;         // color's meaninful name (e.g. ball, goal)
    double merge;       // merge density threshold
    int expected_num;   // expected number of regions (used for merge)
    int l_low,l_high;   // L,A,B component thresholds
    int a_low,a_high;
    int b_low,b_high;
  };

  struct point{
    double x,y;
  };

  struct line{
    point a,b;
  };

  struct rectangle{
    int x,y,w,h;
  };

protected:
  unsigned l_class[CMV_COLOR_LEVELS];
  unsigned a_class[CMV_COLOR_LEVELS];
  unsigned b_class[CMV_COLOR_LEVELS];

  region region_table[CMV_MAX_REGIONS];
  region *region_list[CMV_MAX_COLORS];
  int region_count[CMV_MAX_COLORS];

  rle rmap[CMV_MAX_RUNS];

  color_info colors[CMV_MAX_COLORS];
  int width,height;
  unsigned *map;

  unsigned options;

protected:
// Private functions
  void classifyFrame(image_pixel * restrict img,unsigned * restrict map);
  int  encodeRuns(rle * restrict out,unsigned * restrict map);
  void connectComponents(rle * restrict map,int num);
  int  extractRegions(region * restrict reg,rle * restrict rmap,int num);
  void calcAverageColors(region * restrict reg,int num_reg,
                         image_pixel * restrict img,
                         rle * restrict rmap,int num_runs);
  int  separateRegions(region * restrict reg,int num);
  region *sortRegionListByArea(region * restrict list,int passes);
  void sortRegions(int max_area);

  // density based merging support
  int mergeRegions(region *p,int num,double density_thresh);
  int mergeRegions();

  void clear();

public:
  CMVision()  {clear();}
  ~CMVision() {close();}

  bool initialize(int nwidth,int nheight);
  bool loadOptions(const char *filename);
  bool saveOptions(char *filename);
  bool enable(unsigned opt);
  bool disable(unsigned opt);
  void close();

  bool testClassify(rgb * restrict out,image_pixel * restrict image);
  bool getThreshold(int color,
         int &l_low,int &l_high,
	 int &a_low,int &a_high,
         int &b_low,int &b_high);
  bool setThreshold(int color,
         int l_low,int l_high,
	 int a_low,int a_high,
         int b_low,int b_high);

  unsigned *getMap()
    {return(map);}

  char *getColorName(int color)
    {return(colors[color].name);}
  rgb getColorVisual(int color)
    {return(colors[color].color);}

  color_info *getColorInfo(int color)
    {return(&colors[color]);}
  void getColorInfo(int color,color_info &info)
    {info = colors[color];}
  void setColorInfo(int color,color_info &info)
    {colors[color] = info;}

  bool processFrame(image_pixel *image);
  bool processFrame(unsigned *map);
  int numRegions(int color_id);
  region *getRegions(int color_id);
};

#endif


src/color_gui.h
/*********************************************************************
 * Copyright (c) 2008, Willow Garage, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of Willow Garage, Inc. nor the names of its
 *       contributors may be used to endorse or promote products derived from
 *       this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/

/** \author Nate Koenig and Peter Pastor */

#include <ros/ros.h>
#include <opencv/cv.h>
#include <sensor_msgs/Image.h>
#include <wx/wx.h>

#include "cmvision.h"

class ColorGuiFrame : public wxFrame
{
  enum {ID_Reset=1};

  /// \brief Constructor
  public: ColorGuiFrame();

  /// \brief Quit callback
  public: void OnQuit(wxCommandEvent &event);

  /// \brief Reset callback
  public: void OnReset(wxCommandEvent &event);

  /// \brief On image click callback
  public: void OnClick(wxMouseEvent &event);

  /// \brief Callback for zooming
  public: void OnMouseWheel(wxMouseEvent &event);

  /// \brief Draw an image frame
  public: void DrawImage(const sensor_msgs::ImageConstPtr& msg);

  private: int width_, height_;
  private: wxTextCtrl *abText_;
  private: wxTextCtrl *rgbText_;
  private: wxPanel *image_panel_;

  private: unsigned char *rgb_image_;
  private: unsigned char *lab_image_;

  private: int scale_pos_x_, scale_pos_y_;
  private: float scale_;

  private: CMVision *vision_;
};


class ColorGuiApp : public wxApp
{
  /// \brief On init of the application
  public: bool OnInit();

  /// \brief On update, used for ros::spin
  public: void OnUpdate( wxTimerEvent &event );

  /// \brief The image callback
  private: void imageCB(const sensor_msgs::ImageConstPtr& msg);

  private: ros::Subscriber image_subscriber_;
  private: wxTimer *update_timer_;
  private: ColorGuiFrame *frame_;
};

DECLARE_APP(ColorGuiApp)
IMPLEMENT_APP(ColorGuiApp)


src/color_gui.cpp
/*********************************************************************
 * Copyright (c) 2008, Willow Garage, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of Willow Garage, Inc. nor the names of its
 *       contributors may be used to endorse or promote products derived from
 *       this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/

/** \author Nate Koenig and Peter Pastor */

#include <string>
#include <cv_bridge/cv_bridge.h>

#include "color_gui.h"
#include "conversions.h"

bool ColorGuiApp::OnInit()
{
  char **local_argv = new char*[ argc ];
  for (int i =0; i < argc; i++)
    local_argv[i] = strdup( wxString( argv[i] ).mb_str() );

  ros::init(argc, local_argv, "cmvision");
  ros::NodeHandle node_handle;

  // Subscribe to an image stream
  image_subscriber_ = node_handle.subscribe("image", 1, &ColorGuiApp::imageCB, this);

  frame_ = new ColorGuiFrame();
  frame_->Show(true);
  SetTopWindow(frame_);

  update_timer_ = new wxTimer(this);
  update_timer_->Start( 33 );

  Connect( update_timer_->GetId(), wxEVT_TIMER, wxTimerEventHandler( ColorGuiApp::OnUpdate ), NULL, this);

  return true;
}

void ColorGuiApp::OnUpdate( wxTimerEvent &event )
{
  ros::spinOnce();
}

void ColorGuiApp::imageCB(const sensor_msgs::ImageConstPtr& msg)
{
  frame_->DrawImage( msg );
}




ColorGuiFrame::ColorGuiFrame()
  : wxFrame(NULL, -1, wxT("Color Gui"), wxDefaultPosition, wxSize(800,600), wxDEFAULT_FRAME_STYLE)
{

  wxInitAllImageHandlers();

  wxMenuBar *menuBar = new wxMenuBar;
  wxMenu *file_menu = new wxMenu;

  wxMenuItem *item = file_menu->Append(ID_Reset, wxT("&Reset\tCtrl-R"));
  Connect(item->GetId(), wxEVT_COMMAND_MENU_SELECTED, wxCommandEventHandler(ColorGuiFrame::OnReset), NULL, this);

  item = file_menu->Append(wxID_EXIT, wxT("&Quit\tCtrl-Q"));
  Connect(item->GetId(), wxEVT_COMMAND_MENU_SELECTED, wxCommandEventHandler(ColorGuiFrame::OnQuit), NULL, this);

  menuBar->Append( file_menu, _("&File") );
  
  SetMenuBar( menuBar );

  CreateStatusBar();
  SetStatusText( _("Click on the image to select colors. Use scroll wheel to zoom.") );

  image_panel_ = new wxPanel(this, wxID_ANY, wxPoint(0,0), wxSize(640,480));

  wxStaticText *rgblabel = new wxStaticText(this, -1, wxT("RGB:"));
  rgbText_ = new wxTextCtrl(this,-1,wxT(""));

  wxStaticText *abLabel = new wxStaticText(this, -1, wxT("AB:"));
  abText_ = new wxTextCtrl(this,-1,wxT(""));

  wxBoxSizer *hsizer1 = new wxBoxSizer(wxHORIZONTAL);
  wxBoxSizer *hsizer2 = new wxBoxSizer(wxHORIZONTAL);
  wxBoxSizer *vsizer = new wxBoxSizer(wxVERTICAL);

  hsizer2->Add( rgblabel, 0, wxALIGN_CENTER_VERTICAL| wxLEFT, 10);
  hsizer2->Add( rgbText_, 1, wxALIGN_CENTER_VERTICAL | wxLEFT, 2);

  hsizer2->Add( abLabel, 0, wxALIGN_CENTER_VERTICAL | wxLEFT, 10);
  hsizer2->Add( abText_, 1, wxALIGN_CENTER_VERTICAL | wxLEFT, 2);
  hsizer2->Add(50,1,0);

  vsizer->Add(image_panel_, 0, wxEXPAND);
  vsizer->Add(20,2,0);
  vsizer->Add(hsizer2, 0, wxALIGN_LEFT | wxEXPAND );
  this->SetSizer(vsizer);

	width_ = 0;
	height_ = 0;

  scale_ = 1.0;
  width_ = 0;
  height_ = 0;

  rgb_image_ = NULL;
  lab_image_ = NULL;

  vision_ = new CMVision();

  image_panel_->Connect(wxEVT_LEFT_UP, wxMouseEventHandler(ColorGuiFrame::OnClick), NULL, this);

  Connect(wxEVT_MOUSEWHEEL, wxMouseEventHandler(ColorGuiFrame::OnMouseWheel));

}

void ColorGuiFrame::OnQuit(wxCommandEvent &event)
{
  Close(true);
}

void ColorGuiFrame::DrawImage(const sensor_msgs::ImageConstPtr& msg)
{
  IplImage cvImageRef, *cvImage;
  CvSize size;

  const sensor_msgs::Image img = *msg;

  // Get the image as and RGB image
  cv_bridge::CvImagePtr image_ptr = cv_bridge::toCvCopy(msg, "rgb8");
  cvImageRef = IplImage(image_ptr->image);
  cvImage = &cvImageRef;

  size = cvGetSize(cvImage);

  if (width_ != size.width || height_ != size.height)
  {
    if (!rgb_image_)
      delete[] rgb_image_;
    rgb_image_ = new unsigned char[size.width * size.height * 3];

    if (!lab_image_)
      delete[] lab_image_;
    lab_image_ = new unsigned char[size.width * size.height * 3];

    if (!(vision_->initialize(size.width, size.height)))
    {
      width_ = height_ = 0;
      ROS_ERROR("Vision init failed.");
      return;
    }
  }

  width_ = size.width;
  height_ = size.height;

  memcpy(rgb_image_, cvImage->imageData, width_ * height_ * 3);

  // Convert image to LAB color space
  cvCvtColor(cvImage, cvImage, CV_RGB2Lab);
  
  memcpy(lab_image_, cvImage->imageData, width_ * height_ * 3);

  // Find the color blobs
  if (!vision_->processFrame(reinterpret_cast<image_pixel*> (cvImage->imageData)))
  {
	ROS_ERROR("Frame error.");
	return;
  }

  int xsrc = (scale_pos_x_*scale_) - scale_pos_x_;
  int ysrc = (scale_pos_y_*scale_) - scale_pos_y_;

  wxImage image(width_, height_, rgb_image_, true);
  image.Rescale(width_*scale_,height_*scale_);

  wxBitmap bitmap(image);

  wxMemoryDC memDC;
  memDC.SelectObject(bitmap);

  wxClientDC dc(image_panel_);
  if (xsrc < 0 || ysrc < 0)
    dc.Clear();

  dc.Blit(0,0, 640, 480, &memDC, xsrc, ysrc);

	// Get all the blobs
	for (int ch = 0; ch < CMV_MAX_COLORS; ++ch)
	{
		// Get the descriptive color
		rgb c = vision_->getColorVisual(ch);

		// Grab the regions for this color
		CMVision::region* r = NULL;

		for (r = vision_->getRegions(ch); r != NULL; r = r->next)
		{
      dc.SetBrush(*wxTRANSPARENT_BRUSH);
      int x1 = (r->x1*scale_) - xsrc;
      int y1 = (r->y1*scale_) - ysrc;
      int x2 = (r->x2*scale_) - xsrc;
      int y2 = (r->y2*scale_) - ysrc;

      int w = x2 - x1;
      int h = y2 - y1;

      dc.DrawRectangle(x1, y1, w, h);
		}
	}

  int x, y;
  GetPosition(&x, &y);
// Setting size is commented out because it breaks the gui:
// Text boxes become invisible.
//  SetSize(x,y, width_, height_+80);
}


void ColorGuiFrame::OnReset(wxCommandEvent &event)
{
  vision_->setThreshold(0, 0, 0, 0, 0);
  rgbText_->SetValue(wxString::FromAscii(""));
  abText_->SetValue(wxString::FromAscii(""));
}

void ColorGuiFrame::OnClick(wxMouseEvent &event)
{
  int r, g, b, l, a, b2;

  int px = (event.m_x/scale_) + ((scale_pos_x_*scale_) - scale_pos_x_)/scale_;
  int py = (event.m_y/scale_) + ((scale_pos_y_*scale_) - scale_pos_y_)/scale_;

  r = rgb_image_[py * (width_ * 3) + px * 3 + 0];
  g = rgb_image_[py * (width_ * 3) + px * 3 + 1];
  b = rgb_image_[py * (width_ * 3) + px * 3 + 2];

  std::ostringstream stream1;
  stream1 << "(" <<  r  << ", " << g << ", " << b << ")";
  rgbText_->SetValue(wxString::FromAscii(stream1.str().c_str()));

  l = lab_image_[py * (width_ * 3) + px * 3 + 0];
  a = lab_image_[py * (width_ * 3) + px * 3 + 1];
  b2 = lab_image_[py * (width_ * 3) + px * 3 + 2];


  int l_low, l_high, a_low, a_high, b_low, b_high;

  vision_->getThreshold(0, a_low, a_high, b_low, b_high);

  if (a_low == 0 && a_high == 0)
  {
    a_low = a;
    a_high = a;
  }
  if (b_low == 0 && b_high == 0)
  {
    b_low = b2;
    b_high = b2;
  }

  a_low = std::min(a, a_low);
  a_high = std::max(a, a_high);

  b_low = std::min(b2, b_low);
  b_high = std::max(b2, b_high);

  vision_->setThreshold(0, a_low, a_high, b_low, b_high);

  std::ostringstream stream;
  stream << "( " << a_low << ":" << a_high << ", "
         << b_low << ":" << b_high << " ) ";

  abText_->SetValue(wxString::FromAscii(stream.str().c_str()));
}

void ColorGuiFrame::OnMouseWheel(wxMouseEvent &event)
{
  if (event.GetWheelRotation() < 0)
    scale_ *= 0.9;
  else
    scale_ *= 1.1;

  scale_pos_x_ = event.m_x;
  scale_pos_y_ = event.m_y;
}


src/cmvision_color_blob_finder.h
/*
 *  Player - One Hell of a Robot Server
 *  Copyright (C) 2000
 *     Brian Gerkey, Kasper Stoy, Richard Vaughan, & Andrew Howard
 *     Andrew Martignoni III
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
/*
 * Uses CMVision to retrieve the blob data
 */
// author Andy Martignoni III, Brian Gerkey, Brendan Burns, Ben Grocholsky, Brad Kratochvil

#include <ros/ros.h>
#include <sensor_msgs/Image.h>

#include <opencv/cv.h>

#include <cmvision/Blobs.h>
#include "conversions.h"
#include "cmvision.h"
#include "capture.h"

#define CMV_NUM_CHANNELS CMV_MAX_COLORS
#define CMV_HEADER_SIZE 4*CMV_NUM_CHANNELS
#define CMV_BLOB_SIZE 16
#define CMV_MAX_BLOBS_PER_CHANNEL 10

#define DEFAULT_CMV_WIDTH CMV_DEFAULT_WIDTH
#define DEFAULT_CMV_HEIGHT CMV_DEFAULT_HEIGHT

namespace color_blob_track
{

class CMVisionColorBlobFinder
{
public:

	/// \brief Constructor
	CMVisionColorBlobFinder();

	/// \brief Destructor
	virtual ~CMVisionColorBlobFinder();

	/*! \brief initialization function
	 * @param node_handle
	 * @return true if initialization successful, false else.
	 */
	bool initialize(ros::NodeHandle &node_handle);

	/// \brief Image callback
	void imageCB(const sensor_msgs::ImageConstPtr& msg);

private:

	ros::Publisher blob_publisher_;
	ros::Subscriber image_subscriber_;

	// sensor_msgs::Image image;

	bool debug_on_;
	uint16_t width_;
	uint16_t height_;
	std::string color_filename_;
	uint8_t *lab_image_;

	unsigned int blob_count_;

	CMVision *vision_;

	cmvision::Blobs blob_message_;

	bool mean_shift_on_;
	double spatial_radius_;
	double color_radius_;

};

}


src/cmvision_color_blob_finder.cpp
/*
 *  Player - One Hell of a Robot Server
 *  Copyright (C) 2000
 *     Brian Gerkey, Kasper Stoy, Richard Vaughan, & Andrew Howard
 *     Andrew Martignoni III
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

/*
 * Uses CMVision to retrieve the blob data
 */
// author Andy Martignoni III, Brian Gerkey, Brendan Burns, Ben Grocholsky, Brad Kratochvil

#include <opencv/highgui.h>
#include <ros/time.h>
#include <cv_bridge/cv_bridge.h>

#include "cmvision_color_blob_finder.h"

using namespace color_blob_track;

CMVisionColorBlobFinder::CMVisionColorBlobFinder() :
	debug_on_(false), width_(0), height_(0), color_filename_(""), lab_image_(NULL), blob_count_(0), vision_(NULL), mean_shift_on_(false), spatial_radius_(0),
			color_radius_(0)
{
}

CMVisionColorBlobFinder::~CMVisionColorBlobFinder()
{
	if (vision_)
	{
		delete vision_;
	}
}

bool CMVisionColorBlobFinder::initialize(ros::NodeHandle &node_handle)
{

	lab_image_ = NULL;
	width_ = 0;
	height_ = 0;

	vision_ = new CMVision();

	blob_count_ = 0;
	blob_message_.blob_count = 0;

	// Get the color file. This defines what colors to track
	if (!node_handle.getParam("/cmvision/color_file", color_filename_))
	{
		ROS_ERROR("Could not find color calibration file name \"/cmvision/color_file\" in namespace: %s.", node_handle.getNamespace().c_str());
		return false;
	}

	// Get the level of debug output
	node_handle.param("/cmvision/debug_on", debug_on_, false);

	// check whether mean shift is turned on
	if (!node_handle.getParam("/cmvision/mean_shift_on", mean_shift_on_))
	{
		ROS_ERROR("Could not find mean shift flag \"/cmvision/mean_shift_on\" in namespace: %s.", node_handle.getNamespace().c_str());
		return false;
	}
	else
	{
		if (!node_handle.getParam("/cmvision/spatial_radius_pix", spatial_radius_))
		{
			ROS_ERROR("Could not get spatial_radius_pix from param server \"/cmvision/spatial_radius_pix\" in namespace: %s.", node_handle.getNamespace().c_str());
			return false;
		}

		if (!node_handle.getParam("/cmvision/color_radius_pix", color_radius_))
		{
			ROS_ERROR("Could not get color_radius_pix from param server \"/cmvision/color_radius_pix\" in namespace: %s.", node_handle.getNamespace().c_str());
			return false;
		}
	}

	// Subscribe to an image stream
	image_subscriber_ = node_handle.subscribe("image", 1, &CMVisionColorBlobFinder::imageCB, this);

	// Advertise our blobs
	blob_publisher_ = node_handle.advertise<cmvision::Blobs> ("blobs", 1);

	if (debug_on_)
	{
		cvNamedWindow("Image");
	}

	return true;
}

void CMVisionColorBlobFinder::imageCB(const sensor_msgs::ImageConstPtr& msg)
{
	IplImage cvImageRef, *cvImage, *debug_img_;
	CvSize size;

	const sensor_msgs::Image img = *msg;

	// Timing
	// struct timeval timeofday;
	// gettimeofday(&timeofday,NULL);
	// ros::Time startt = ros::Time().fromNSec(1e9*timeofday.tv_sec + 1e3*timeofday.tv_usec);
	ros::WallTime startt = ros::WallTime::now();

	// Get the image as and RGB image
        cv_bridge::CvImagePtr image = cv_bridge::toCvCopy(msg, "bgr8");
        cvImageRef = IplImage(image->image);
        cvImage = &cvImageRef;
        
	size = cvGetSize(cvImage);

	// this shouldn't change often
	if ((size.width != width_) || (size.height != height_))
	{
		if (!(vision_->initialize(size.width, size.height)))
		{
			width_ = height_ = 0;
			ROS_ERROR("Vision init failed.");
			return;
		}

		if (!color_filename_.empty())
		{
			if (!vision_->loadOptions(color_filename_.c_str()))
			{
				ROS_ERROR("Error loading color file");
				return;
			}
		}
		else
		{
			ROS_ERROR("No color file given.  Use the \"mColorFile\" "
				"option in the configuration file.");
			return;
		}

		width_ = size.width;
		height_ = size.height;

		blob_message_.image_width = size.width;
		blob_message_.image_height = size.height;

		if (lab_image_)
		{
			delete[] lab_image_;
		}
		lab_image_ = new uint8_t[width_ * height_ * 3];

	}

	// Smooth the image, if turned on
	if (mean_shift_on_)
	{
		cvPyrMeanShiftFiltering(cvImage, cvImage, spatial_radius_, color_radius_);
	}

	if (debug_on_) {
		debug_img_ = cvCloneImage(cvImage);
	}

	cvCvtColor(cvImage, cvImage, CV_BGR2Lab);

	// Find the color blobs
	if (!vision_->processFrame(reinterpret_cast<image_pixel*> (cvImage->imageData)))
	{
		ROS_ERROR("Frame error.");
		return;
	}

  //image_bridge_.fromImage(*msg, "bgr8");
  //cvImage = image_bridge_.toIpl();

	// Get all the blobs
	blob_count_ = 0;
	for (int ch = 0; ch < CMV_MAX_COLORS; ++ch)
	{
		// Get the descriptive color
		rgb c = vision_->getColorVisual(ch);
		char* name = vision_->getColorName(ch);

		// Grab the regions for this color
		CMVision::region* r = NULL;

		for (r = vision_->getRegions(ch); r != NULL; r = r->next)
		{
			// Resize the blob message
			if (blob_count_ >= blob_message_.blobs.size())
			{
				blob_message_.blobs.resize(blob_message_.blobs.size() + 1);
			}

			if (debug_on_)
			{
				cvRectangle(debug_img_, cvPoint(r->x1, r->y1), cvPoint(r->x2, r->y2), CV_RGB(c.red, c.green, c.blue));
			}

			blob_message_.blobs[blob_count_].name = name;
			blob_message_.blobs[blob_count_].red = c.red;
			blob_message_.blobs[blob_count_].green = c.green;
			blob_message_.blobs[blob_count_].blue = c.blue;
			blob_message_.blobs[blob_count_].area = r->area;
			blob_message_.blobs[blob_count_].x = rint(r->cen_x + .5);
			blob_message_.blobs[blob_count_].y = rint(r->cen_y + .5);
			blob_message_.blobs[blob_count_].left = r->x1;
			blob_message_.blobs[blob_count_].right = r->x2;
			blob_message_.blobs[blob_count_].top = r->y1;
			blob_message_.blobs[blob_count_].bottom = r->y2;

			blob_count_++;
		}
	}

  if (blob_count_ < blob_message_.blobs.size())
    blob_message_.blobs.resize(blob_count_);

	if (debug_on_)
	{
		cvShowImage("Image", debug_img_);

		cvWaitKey(3);

		cvReleaseImage(&debug_img_);
	}

	blob_message_.blob_count = blob_count_;

	blob_message_.header.stamp = ros::Time::now();

	blob_publisher_.publish(blob_message_);

	// Timing
	// gettimeofday(&timeofday,NULL);
	// ros::Time endt = ros::Time().fromNSec(1e9*timeofday.tv_sec + 1e3*timeofday.tv_usec);
	ros::WallTime endt = ros::WallTime::now();
	ros::WallDuration diff = endt - startt;
	ROS_DEBUG_STREAM_NAMED("cmvision", "Color Blob Detection duration " << diff.toSec());
}
